File: packages/ts/octospaces-sdk/src/sync/pull-cache.ts
/**
 * The app's offline-first read cache for every {@link StarfishClient}.
 *
 * Backs the SDK's {@link PullCache} (read-through pull cache) with the kv layer
 * (localStorage on web, AsyncStorage on native). When a client is built with this
 * cache, every successful structured `pull()` is written through, and a pull that
 * fails because the transport is unreachable falls back to the last-synced snapshot.
 *
 * SECURITY: the SDK caches the RAW server response only. For E2E collections that
 * payload is the SEALED ciphertext the server holds — never the decrypted form —
 * so this cache is ciphertext-at-rest by construction.
 */
import type { PullCache } from '@drakkar.software/starfish-client';

import { kvGet, kvSet } from '../core/adapters.js';

const PREFIX = 'octospaces.pullcache.';

/**
 * Max age for a cached snapshot before it's treated as a miss. Generous (30 days)
 * because for an offline-first app any last-synced data beats none.
 */
export const PULL_CACHE_MAX_AGE_MS = 30 * 24 * 60 * 60 * 1000;

let shared: PullCache | undefined;

/** The shared app-wide pull cache (one instance, reused across every client). */
export function pullCache(): PullCache {
  return (shared ??= {
    get: (key) => kvGet(PREFIX + key),
    set: (key, value) => kvSet(PREFIX + key, value),
  });
}

File: packages/python/octospaces-sdk/octospaces_sdk/sync/pull_cache.py
"""Offline pull cache — persists pulled docs in the KV store.

Mirrors ``packages/ts/octospaces-sdk/src/sync/pull-cache.ts``.
"""

from __future__ import annotations

import json
import time
from typing import Any, Optional

from octospaces_sdk.core.adapters import kv_get, kv_remove, kv_set

PULL_CACHE_MAX_AGE_MS = 30 * 24 * 60 * 60 * 1000  # 30 days
_PREFIX = "octospaces.pullcache."


async def cache_pull(path: str, doc: Any) -> None:
    entry = {"doc": doc, "ts": int(time.time() * 1000)}
    await kv_set(f"{_PREFIX}{path}", json.dumps(entry))


async def load_cached_pull(path: str) -> Optional[Any]:
    raw = await kv_get(f"{_PREFIX}{path}")
    if raw is None:
        return None
    try:
        entry = json.loads(raw)
    except json.JSONDecodeError:
        return None
    age = int(time.time() * 1000) - entry.get("ts", 0)
    if age > PULL_CACHE_MAX_AGE_MS:
        await kv_remove(f"{_PREFIX}{path}")
        return None
    return entry["doc"]
